samples/azure-quantum/entanglement-swap/ES-quantinuum-qsharp.ipynb

```diff
@@ -0,0 +1,1120 @@
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "%azure.target quantinuum.hqs-lt-s2-sim"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "%azure.submit TeleportFidelityHelper jobName=\"Teleporation fidelity simulation\""
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Now we will look into our job and wait for its completion using `%azure.status` and once complete see results using `%azure.output`."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "%azure.status"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "%azure.output"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Let us now enhance our code by building entanglement swapping. The idea of entanglement swapping is that if Alice and Bob do not have a direct connection but connections to third parties such that a path exists, then each pair of adjacent parties can create Bell states and teleport Alice's second qubit along this path until it reaches Bob.\n",
+    "\n",
+    "First we build the base case of entanglement swapping `EntanglementSwap3`, then we proceed to build the general case `EntanglementSwapN`\n",
+    "\n",
+    "We will test entanglement swapping with 3 parties by using the Bell pair we create to send a specific qubit state. With the advanced debugging tools of Q# we can introspect that states mid-simulation allowing us to perform this test. To validate that teleportation protocol was implemented correctly, we will use assertion features provided by Q#. Later in this notebook we will test entanglement swapping by checking that Alice and Bob have a Bell state since we will run on real hardware (or accurate simulators thereof) which does not allow such introspection. Validating concepts on small scale using Q# debugging features before proceeding to building full-scale algorithms is a good quantum software development practice."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 10,
+   "metadata": {},
+   "outputs": [
+    {
+     "data": {
+      "application/x-qsharp-data": "[\"EntanglementSwap3\",\"TestEntanglementSwap3\"]",
+      "text/html": [
+       "<ul><li>EntanglementSwap3</li><li>TestEntanglementSwap3</li></ul>"
+      ],
+      "text/plain": [
+       "EntanglementSwap3, TestEntanglementSwap3"
+      ]
+     },
+     "execution_count": 10,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "operation EntanglementSwap3(alice_charlie : Qubit[], charlie_bob : Qubit[]) : (Qubit, Qubit) {\n",
```

Comment:
Since the number of qubits is guaranteed to be exactly two here, it may be more readable to make that explicit by using `aliceCharlie : (Qubit, Qubit)` instead of `Qubit[]` in this case. You can then later just call `CreateBellState(aliceCharlie)` by using the tuple-in–tuple-out semantics of Q#.

Reply:
Could you link to some more documentation on those semantics? I don't think I have seen those before

Reply:
It should be on the Q# user guide, I'll find a link shortly. In the meantime, though, every callable in Q# takes exactly one input and returns exactly one output. In particular, `CreateBellState` has type `(Qubit, Qubit) => Unit is Adj + Ctl`, such that you can pass a single value of type `(Qubit, Qubit)` to that operation. That works in general alongside singleton–tuple equivalence, which says that `'T` and `('T)` are exactly the same type for all `'T`.

Reply:
https://docs.microsoft.com/en-us/azure/quantum/user-guide/language/typesystem/singletontupleequivalence, don't have a link handy for TI-TO semantics, sorry.